### PR TITLE
feat(forge): Implement a very very bad, but "functional" agent by default

### PR DESCRIPTION
**Pull request by ntindle**
### Background

We’ve received some notes about Forge being difficult to use and a bit broken. This PR aims to solve some of the core issues with that. 

### Changes 🏗️

- Fixes various issues
  - Windows Compatibility
  - FastAPI interface Issues
  - Hardcoded paths removed
  - Formatting
  - Ability -> Action Fixes
   - Prompt
   - Imports
   - Actions Files themselves
  - Notes for the codebase
  - Endpoint Error message fixes
- Updates typing throughout
- Swaps agent from one that always replies Washington DC to one that has a very bad agent loop

Do we want to keep Washington DCcand make a real agent instead?
 

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [ ] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [ ] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [ ] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [ ] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

### CODE REVIEW

File: autogpts/forge/forge/sdk/model.py [hunk 1]
```diff
@@ -12,51 +12,42 @@
...
+    current_page: int = Field(description="Current_page page number.", example=1)
+    page_size: int = Field(description="Number of items per page.", example=25)
 
 
 class Artifact(BaseModel):
     created_at: datetime = Field(
-        ...,
         description="The creation datetime of the task.",
         example="2023-01-01T00:00:00Z",
-        json_encoders={datetime: lambda v: v.isoformat()},
```
Comment (ntindle) [inline autogpts/forge/forge/sdk/model.py:35]: This is why the openapi docs doesn't work btw

File: autogpts/forge/forge/sdk/prompting.py [hunk 2]
```diff
@@ -108,6 +108,7 @@ def load_prompt(self, template: str, **kwargs) -> str:
             template = os.path.join(self.model, template)
             if self.debug_enabled:
                 LOG.debug(f"Loading template: {template}")
+            template = template.replace("\\", "/")
```
Comment (ntindle) [inline autogpts/forge/forge/sdk/prompting.py]: windows compat

File: autogpts/forge/forge/actions/file_system/files.py [hunk 3]
```diff
@@ -1,22 +1,24 @@
...
     name="list_files",
     description="List files in a directory",
     parameters=[
-        {
-            "name": "path",
-            "description": "Path to the directory",
-            "type": "string",
-            "required": True,
-        }
+        ActionParameter(
```
Comment (ntindle) [inline autogpts/forge/forge/actions/file_system/files.py:12]: type fix

File: autogpts/forge/forge/actions/file_system/files.py [hunk 4]
```diff
@@ -1,22 +1,24 @@
...
+            name="path",
+            description="Path to the directory",
+            type="string",
+            required=True,
+        )
     ],
     output_type="list[str]",
 )
-async def list_files(agent, task_id: str, path: str) -> List[str]:
+async def list_files(agent: Agent, task_id: str, path: str) -> List[str]:
```
Comment (ntindle) [inline autogpts/forge/forge/actions/file_system/files.py:21]: type hint

File: autogpts/forge/forge/actions/registry.py [hunk 5]
```diff
@@ -186,8 +188,11 @@ async def run_action(
 
 if __name__ == "__main__":
     import sys
+    import asyncio
```
Comment (ntindle) [inline autogpts/forge/forge/actions/registry.py:190]: swifty removal and replacement with more amenable action to not actually having an agent seeded. list_files requires a full agent to be able to be pulled in to actually work

File: autogpts/forge/forge/actions/web/web_selenium.py [hunk 6]
```diff
@@ -231,7 +233,9 @@ async def read_webpage(
     except WebDriverException as e:
         # These errors are often quite long and include lots of context.
         # Just grab the first line.
-        msg = e.msg.split("\n")[0]
+        msg = "An error occurred while trying to load the page"
```
Comment (ntindle) [inline autogpts/forge/forge/actions/web/web_selenium.py:236]: type fix

File: autogpts/forge/forge/agent.py [hunk 7]
```diff
@@ -123,24 +127,84 @@ async def execute_step(self, task_id: str, step_request: StepRequestBody) -> Ste
...
+        # step.output = "Washington D.C"
+
+        # LOG.info(
+        #     f"\t✅ Final Step completed: {step.step_id}. \n"
+        #     + f"Output should be placeholder text Washington D.C. You'll need to \n"
+        #     + f"modify execute_step to include LLM behavior. Follow the tutorial "
+        #     + f"if confused. "
+        # )
+
+        # return step
```
Comment (ntindle) [inline autogpts/forge/forge/agent.py:152]: this is from the tutorial so left here, but replaced with something that "can" work. Its bad at responding though. Could really use some help on the prompts to maek this work sometimes

File: autogpts/forge/forge/sdk/__init__.py [hunk 8]
```diff
@@ -2,7 +2,7 @@
 The Forge SDK. This is the core of the Forge. It contains the agent protocol, which is the
 core of the Forge.
 """
-from ..llm import chat_completion_request, create_embedding_request, transcribe_audio
+# from forge.llm import chat_completion_request, create_embedding_request, transcribe_audio
```
Comment (ntindle) [inline autogpts/forge/forge/sdk/__init__.py:5]: Circular import fix. 

Why do we need this file?

File: autogpts/forge/forge/sdk/agent.py [hunk 9]
```diff
@@ -140,6 +140,9 @@ async def list_steps(
     async def execute_step(self, task_id: str, step_request: StepRequestBody) -> Step:
         """
         Create a step for the task.
+
+        Note: this is a placeholder method and must be overridden by the agent with the actions
+        that the agent should perform on one step. Override this method in the ForgeAgent class.
```
Comment (ntindle) [inline autogpts/forge/forge/sdk/agent.py:145]: Pointer for users of this to look in the right place

File: autogpts/forge/forge/sdk/routes/agent_protocol.py [hunk 10]
```diff
@@ -43,15 +56,15 @@ async def root():
     """
     Root endpoint that returns a welcome message.
     """
-    return Response(content="Welcome to the AutoGPT Forge")
+    return {"message": "Welcome to the AutoGPT Forge"}
```
Comment (ntindle) [inline autogpts/forge/forge/sdk/routes/agent_protocol.py:60]: Type fix. Pydantic will turn into a response automatically

File: autogpts/forge/forge/sdk/routes/agent_protocol.py [hunk 11]
```diff
@@ -81,27 +94,22 @@ async def create_agent_task(request: Request, task_request: TaskRequestBody) ->
                 "artifacts": [],
             }
     """
-    agent = request["agent"]
+    agent: Agent = request["agent"]
```
Comment (ntindle) [inline autogpts/forge/forge/sdk/routes/agent_protocol.py:98]: Type hinting throughout the file

File: autogpts/forge/forge/sdk/routes/agent_protocol.py [hunk 12]
```diff
@@ -81,27 +94,22 @@ async def create_agent_task(request: Request, task_request: TaskRequestBody) ->
...
 
     try:
         task_request = await agent.create_task(task_request)
-        return Response(
-            content=task_request.json(),
-            status_code=200,
-            media_type="application/json",
-        )
+        return task_request
```
Comment (ntindle) [inline autogpts/forge/forge/sdk/routes/agent_protocol.py:102]: pydantic and fastapi will handle the conversions

File: autogpts/forge/forge/sdk/routes/agent_protocol.py [hunk 13]
```diff
@@ -81,27 +94,22 @@ async def create_agent_task(request: Request, task_request: TaskRequestBody) ->
...
     except Exception:
         LOG.exception(f"Error whilst trying to create a task: {task_request}")
-        return Response(
-            content=json.dumps(
+        raise HTTPException(
+            status_code=500,
+            detail=json.dumps(
```
Comment (ntindle) [inline autogpts/forge/forge/sdk/routes/agent_protocol.py:107]: pydantic prefers raise over response

File: autogpts/forge/forge/sdk/routes/agent_protocol.py [hunk 14]
```diff
@@ -267,8 +263,8 @@ async def get_agent_task(request: Request, task_id: str) -> Task:
 async def list_agent_task_steps(
     request: Request,
     task_id: str,
-    page: Optional[int] = Query(1, ge=1),
-    page_size: Optional[int] = Query(10, ge=1, alias="pageSize"),
+    page: Optional[int] = Query(default=1, ge=1),
+    page_size: Optional[int] = Query(default=10, ge=1, alias="pageSize"),
```
Comment (ntindle) [inline autogpts/forge/forge/sdk/routes/agent_protocol.py:268]: type hints

File: autogpts/forge/forge/sdk/routes/agent_protocol.py [hunk 15]
```diff
@@ -304,33 +300,27 @@ async def list_agent_task_steps(
...
-    agent = request["agent"]
+    agent: Agent = request["agent"]
     try:
-        steps = await agent.list_steps(task_id, page, page_size)
-        return Response(
-            content=steps.json(),
-            status_code=200,
-            media_type="application/json",
-        )
+        steps = await agent.list_steps(task_id, page or 1, page_size or 10)
```
Comment (ntindle) [inline autogpts/forge/forge/sdk/routes/agent_protocol.py:306]: type error fix

File: autogpts/forge/forge/sdk/routes/agent_protocol.py [hunk 16]
```diff
@@ -377,38 +367,32 @@ async def execute_agent_task_step(
                 ...
             }
     """
-    agent = request["agent"]
+    agent: Agent = request["agent"]
     try:
         # An empty step request represents a yes to continue command
         if not step:
-            step = StepRequestBody(input="y")
+            step = StepRequestBody(name=None, input="y")
```
Comment (ntindle) [inline autogpts/forge/forge/sdk/routes/agent_protocol.py:375]: type error fix

File: autogpts/forge/forge/sdk/prompting.py (hunk 2 above)
Comment (Pwuts) [reply autogpts/forge/forge/sdk/prompting.py]: Why needed for Windows compat? Do you know of `Path.as_posix()`?

File: autogpts/forge/forge/agent.py [hunk 17]
```diff
@@ -1,14 +1,18 @@
+import json
+import pprint
 from forge.actions import ActionRegister
```
Comment (Pwuts) [inline autogpts/forge/forge/agent.py]: Improper formatting; pls run linters

**Review by Pwuts — DISMISSED** [review]
Looks good overall

**Review by Pwuts — CHANGES_REQUESTED** [review]
Application [crashes when the LLM picks a non-existent command](https://github.com/Significant-Gravitas/AutoGPT/actions/runs/8930811580/job/24531663813?pr=7021#step:5:631), which is also why it fails CI

**Comment by Swiftyos** [comment]
@ntindle closing this now as too much has changed since it was made

**Pull request closed without merging**